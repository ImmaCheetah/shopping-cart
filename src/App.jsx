import { useState, useEffect } from 'react'
import ShopIntro from './components/ShopIntro/ShopIntro';
import PropTypes from 'prop-types';
import './App.css'


const RenderName = (props) => {
  return <div>{props.name}</div>;
};

RenderName.propTypes = {
  name: PropTypes.string.isRequired,
};

const App = () => {
  const [heading, setHeading] = useState("Magnificent Monkeys");

  const clickHandler = () => {
    setHeading("Radical Rhinos");
  };

  return (
    <>
      <ShopIntro />
      <button type="button" onClick={clickHandler}>
        Click Me
      </button>
      <h1>{heading}</h1>
      <ul>
        <li>1</li>
        <li>2</li>
        <li>3</li>
        <li>4</li>
      </ul>
    </>
  );
};

export default App
