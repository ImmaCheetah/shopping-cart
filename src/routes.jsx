import { Children } from "react";
import App from "./App";
import RootLayout from "./components/RootLayout/RootLayout";
import ErrorPage from "./components/ErrorPage/ErrorPage";
// import Profile from "./components/Profile";
// import DefaultProfile from "./components/DefaultProfile";

const routes = [
  {
    path: "/",
    element: <RootLayout />,
    errorElement: <ErrorPage />,
    children: [
        { index: true, element: <App /> },
    ]
  },
];

export default routes;