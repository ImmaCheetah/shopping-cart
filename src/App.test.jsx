import { describe, it, expect, vi } from "vitest";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import { RouterProvider, createMemoryRouter } from "react-router-dom";
import routes from "./routes";

global.fetch = vi.fn(() => {
  const data = [{ id: "1", title: "shirt" }];

  return Promise.resolve({
    json: () => Promise.resolve(data),
  });
});

describe("fetching", () => {
  const router = createMemoryRouter(routes);
  it("shows error message with failed fetch", async () => {
    global.fetch.mockImplementationOnce(() => {
      return Promise.reject({ message: "A network error was encountered" });
    });

    render(
      <RouterProvider router={router}>
        <App />
      </RouterProvider>,
    );
    screen.debug();
    const errorMsg = await screen.findByText("A network error was encountered");

    expect(errorMsg).toBeInTheDocument();
  });

  it("fetches correct data", async () => {
    render(
      <RouterProvider router={router}>
        <App />
      </RouterProvider>,
    );
    const user = userEvent.setup();
    const shopLink = screen.getByTestId('shopLink');

    await user.click(shopLink);
    const fetchData = await screen.findByText("shirt");

    expect(fetchData).toBeInTheDocument();
  });
});

describe("page navigation", () => {
  const router = createMemoryRouter(routes);
  it("should go to shop", async () => {
    render(
      <RouterProvider router={router}>
        <App />
      </RouterProvider>,
    );
    const user = userEvent.setup();

    const welcomeText = screen.getByRole("heading", { name: /welcome/i });
    expect(welcomeText).toBeInTheDocument();

    await user.click(screen.getByTestId('shopLink'));

    expect(screen.getByRole("heading", { name: /our products/i })).toBeInTheDocument();
  });

  it("error page", () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ["/abc"],
    });

    render(<RouterProvider router={router} />);

    expect(
      screen.getByRole("heading", { name: /seems you got lost/i }),
    ).toBeInTheDocument();
  });
});
