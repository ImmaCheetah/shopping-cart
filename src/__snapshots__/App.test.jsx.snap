// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders magnificent monkeys 1`] = `
<div>
  <button
    type="button"
  >
    Click Me
  </button>
  <h1>
    Magnificent Monkeys
  </h1>
  <ul>
    <li>
      1
    </li>
    <li>
      2
    </li>
    <li>
      3
    </li>
    <li>
      4
    </li>
  </ul>
</div>
`;

exports[`toUpperCase 1`] = `"FOOBAR"`;
